components/brave_wallet/browser/ethereum_permission_utils.cc

```diff
@@ -0,0 +1,129 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_wallet/browser/ethereum_permission_utils.h"
+
+#include "base/no_destructor.h"
+#include "base/strings/strcat.h"
+#include "base/strings/string_util.h"
+#include "third_party/re2/src/re2/re2.h"
+#include "url/gurl.h"
+
+namespace {
+
+constexpr char kAddrPattern[] = "addr%3D(0x[[:xdigit:]]{40})";
+
+bool AddAccountToHost(const GURL& old_origin,
+                      const std::string& account,
+                      GURL* new_origin) {
+  if (!old_origin.is_valid() || account.empty() || !new_origin)
+    return false;
+
+  GURL::Replacements replacements;
+  std::string new_host = base::StrCat({old_origin.host_piece(), account});
+  replacements.SetHostStr(new_host);
+
+  *new_origin = old_origin.ReplaceComponents(replacements);
+  return new_origin->is_valid();
+}
+
+void ExtractAddresses(const GURL& origin,
+                      std::queue<std::string>* address_queue) {
+  static const base::NoDestructor<re2::RE2> kAddrRegex(kAddrPattern);
+  DCHECK(origin.is_valid() && address_queue);
+  DCHECK_EQ(origin, origin.GetOrigin());
+
+  re2::StringPiece input(origin.spec());
+  std::string match;
+  while (re2::RE2::FindAndConsume(&input, *kAddrRegex, &match)) {
+    address_queue->push(match);
+  }
+}
+
+bool ParseRequestingOriginInternal(const GURL& origin,
```

Comment:
same as above, please add comments and tests

Reply:
DITTO